### ISSUE TITLE
[C++][Python] Casting map causes pyarrow to crash

### ISSUE THREAD

**Issue by rohanjain101** [apache/arrow#38553]
### Describe the bug, including details regarding any error messages, version, and platform.

```
>>> import pyarrow as pa
>>> z = pa.array([{1: 2}], type=pa.map_(pa.int64(), pa.int64()))
>>> z.cast(pa.map_(pa.int32(), pa.int64()))
arrow/cpp/src/arrow/array/array_nested.cc:431:  Check failed: _s.ok() Operation failed: ValidateChildData(data->child_data)
Bad status: Invalid: Map array keys array should have no nulls
```

### Component(s)

Python

**Comment by felipecrv** [comment]
The fix for this is the same as #41070. Change the `MapArray` validation code in `array_nested.cc` to use `MayHaveNulls()` instead of `arr->null_count != 0`.

**Comment by felipecrv** [comment]
Issue resolved by pull request 41957
https://github.com/apache/arrow/pull/41957
